### PR TITLE
Mat 4393 combine cql editor validation logic

### PR DESCRIPTION
**Pull request by sb-cecilialiu**
## MADiE PR

Jira Ticket: [MAT-4393](https://jira.cms.gov/browse/MAT-4393)
(Optional) Related Tickets:

### All Submissions
* [x] This PR has the JIRA linked.
* [x] Required tests are included
* [x] No extemporaneous files are included (i.e Complied files or testing results)
* [x] This PR is in to the **correct branch**.
* [ ] All Documentation as needed for this PR is Complete (or noted in a TODO or other Ticket)
* [ ] Any breaking changes or failing automation are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [x] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package)
* [x] All CDN/Web dependancies are hosted internally (i.e MADiE-Root Repo)

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
*  The tests appropriately test the new code, including edge cases
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads

### CODE REVIEW

File: src/api/useTerminologyServiceApi.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  errorMsg: string;
+};
+
+export class TerminologyServiceApi {
+  constructor(private getAccessToken: () => string) {}
+
+  async getValueSet(oid: string, locator: string): Promise<FHIRValueSet> {
+    let fhirValueset: FHIRValueSet = null;
+    const baseUrl = await GetServiceUrl();
+    const resp = await axios
```
Comment (RohitKandimalla) [inline src/api/useTerminologyServiceApi.ts]: resp is never used, Do we need to capture it ?

File: src/api/useTerminologyServiceApi.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+      return processCodeSystemErrors(
+        customCqlCodes,
+        "Unable to validate code, Please contact HelpDesk",
+        false
+      );
+    }
+  }
+}
+
+export const GetServiceUrl = async () => {
```
Comment (RohitKandimalla) [inline src/api/useTerminologyServiceApi.ts]: Any specific reason why we are using pascal case ?

File: src/AceEditor/madie-ace-editor.tsx [hunk 3]
```diff
@@ -95,9 +111,51 @@ const MadieAceEditor = ({
...
-      const annotations = mapParserErrorsToAceAnnotations(errors);
-      const aceMarkers = mapParserErrorsToAceMarkers(errors);
+
+      let allErrors: CqlError[] = [];
+      errors.forEach((parseError) => {
+        allErrors.push(parseError);
+      });
+
+      //2. ELM translation errors
+      const translationResults: ElmTranslation = await getElmTranslation(
```
Comment (RohitKandimalla) [inline src/AceEditor/madie-ace-editor.tsx]: Are we doing cql - elm translation and validating terminologies on debouncing?

File: src/validations/editorUtil.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    if (customCqlCodes) {
+      const codeSystemCqlErrors = await ValidateCodes(customCqlCodes);
+      if (codeSystemCqlErrors && codeSystemCqlErrors.length > 0) {
+        codeSystemCqlErrors.map((codeSystem) => {
+          vsacErrorsArray.push(codeSystem);
+        });
+      }
+    }
+  }
+  return vsacErrorsArray;
```
Comment (RohitKandimalla) [inline src/validations/editorUtil.ts]: We could greatly reduce the response time to validate terminology and cql - elm translation. if we can perform async calls to these services at the same time.

File: src/validations/valuesetValidation.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+import useTerminologyServiceApi from "../api/useTerminologyServiceApi";
+
+const validateValueSets = async (
+  valuesetsArray: ElmValueSet[]
+): Promise<ElmTranslationError[]> => {
+  const valuesetsErrors = await GetValueSetErrors(valuesetsArray);
+  return valuesetsErrors;
+};
+
+const GetValueSetErrors = async (
```
Comment (RohitKandimalla) [inline src/validations/valuesetValidation.ts:5]: Usage of Pascal case for a few functions is confusing me, I have seen this for a few other functions as well. Am I missing something?

**Review by RohitKandimalla — CHANGES_REQUESTED** [review]

File: src/AceEditor/madie-ace-editor.tsx (hunk 3 above)
Comment (RohitKandimalla) [reply src/AceEditor/madie-ace-editor.tsx]: I believe that's not what we agreed on, Since for a large CQL, these calls may have a longer response time. So these calls are supposed to happen on a load of the Cql Editor tab and save.

File: src/api/useTerminologyServiceApi.ts (hunk 2 above)
Comment (sb-cecilialiu) [reply src/api/useTerminologyServiceApi.ts]: @RohitKandimalla I had to change it to capital letter as I got lint error about the function name when I checked in the code, it was saying a component function needs to use capital letter and a hook function needs to start with use.

File: src/AceEditor/madie-ace-editor.tsx (hunk 3 above)
Comment (sb-cecilialiu) [reply src/AceEditor/madie-ace-editor.tsx]: @RohitKandimalla Ok so debounce only does parse validation like before, no translation and vsac validation right?

File: src/api/useTerminologyServiceApi.ts (hunk 2 above)
Comment (RohitKandimalla) [reply src/api/useTerminologyServiceApi.ts]: but it's not a React component, it is just a ts file. Maybe we need to revisit the lining rules.

File: src/AceEditor/madie-ace-editor.tsx (hunk 3 above)
Comment (RohitKandimalla) [reply src/AceEditor/madie-ace-editor.tsx]: Yes, Unless there is a conversation that I missed.

File: src/api/useTerminologyServiceApi.ts (hunk 2 above)
Comment (RohitKandimalla) [reply src/api/useTerminologyServiceApi.ts]: My bad, the lint error makes sense, Since it is a custom React hook, we can directly use useTerminologyServiceApi to set the service config while invoking the constructor.

File: src/AceEditor/madie-ace-editor.tsx (hunk 3 above)
Comment (sb-cecilialiu) [reply src/AceEditor/madie-ace-editor.tsx]: @RohitKandimalla Ok got it thanks!

File: src/AceEditor/madie-ace-editor.tsx [hunk 6]
```diff
@@ -198,4 +214,38 @@ const MadieAceEditor = ({
...
+//   await terminologyServiceApi
+//     .checkLogin()
+//     .then(() => {
+//       isLoggedIn = true;
+//     })
+//     .catch((err) => {
+//       isLoggedIn = false;
+//     });
+//   return isLoggedIn;
+// };
```
Comment (RohitKandimalla) [inline src/AceEditor/madie-ace-editor.tsx]: This is no longer required?

File: src/validations/elmTranslateValidation.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import React from "react";
+import useElmTranslationServiceApi, {
+  ElmTranslation,
+} from "../api/useElmTranslationServiceApi";
+import * as _ from "lodash";
```
Comment (RohitKandimalla) [inline src/validations/elmTranslateValidation.ts:5]: Unused imports

File: src/validations/elmTranslateValidation.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+const useTranslateCql = async (cql: string): Promise<ElmTranslation> => {
+  const elmTranslationServiceApi = useElmTranslationServiceApi();
+  if (cql && cql.trim().length > 0) {
+    const [translationResults] = await Promise.all([
+      await elmTranslationServiceApi.translateCqlToElm(cql),
+    ]);
+    return translationResults;
+  } else {
+  }
```
Comment (RohitKandimalla) [inline src/validations/elmTranslateValidation.ts:13]: Missing the else part?
Also, Do we require to have promise.all, when there is only a single promise that we are waiting for.
In addition, we are not using translationResult in this method, instead, the result is just returned back. So I don't think we need to await.

File: src/AceEditor/madie-ace-editor.tsx [hunk 9]
```diff
@@ -198,4 +214,38 @@ const MadieAceEditor = ({
...
+  loggedInUMLS: boolean
+): Promise<ElmTranslationError[]> => {
+  return await useValidateCustomeCqlCodes(customCqlCodes, loggedInUMLS);
+};
+
+export const validateValueSets = async (
+  valuesetsArray: ElmValueSet[],
+  loggedInUMLS: boolean
+): Promise<ElmTranslationError[]> => {
+  return await getValueSetErrors(valuesetsArray, loggedInUMLS);
```
Comment (RohitKandimalla) [inline src/AceEditor/madie-ace-editor.tsx]: await is not required, Since a promise is being returned. Applies for all 3 functions.

File: src/validations/codesystemValidation.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  loggedInUMLS: boolean
+): Promise<ElmTranslationError[]> => {
+  const terminologyServiceApi = useTerminologyServiceApi();
+  const [validatedCodes] = await Promise.all([
+    await terminologyServiceApi.validateCodes(customCqlCodes, loggedInUMLS),
+  ]);
+  const codeSystemCqlErrors: ElmTranslationError[] =
+    mapCodeSystemErrorsToTranslationErrors(validatedCodes);
+  return codeSystemCqlErrors;
+};
```
Comment (RohitKandimalla) [inline src/validations/codesystemValidation.ts:77]: Looks like the function name has a typo.
Promise.all is not required.

**Review by RohitKandimalla — DISMISSED** [review]

**Review by RohitKandimalla — APPROVED** [review]
